Sources/SwiftSyntax/Raw/RawSyntax.swift

```diff
@@ -267,6 +267,50 @@ extension RawSyntax {
   }
 }
 
+extension RawSyntax {
+  /// Enumerate all of the syntax text present in this node, and all
+  /// of its children, to give a source-accurate view of the bytes.
+  ///
+  /// Unlike `description`, this provides a source-accurate representation
+  /// even in the presence of malformed UTF-8 in the input source.
+  ///
+  /// The `SyntaxText` arguments passed to the visitor are only guaranteed
+  /// to be valid within that call. It is unsafe to escape the `SyntaxValue`
+  /// values outside of the closure.
+  public func withEachSyntaxText(body: (SyntaxText) throws -> Void) rethrows {
+    switch rawData.payload {
+    case .parsedToken(let dat):
+      if dat.presence == .present {
+        try body(dat.wholeText)
+      }
+    case .materializedToken(let dat):
+      if dat.presence == .present {
+        for p in dat.leadingTrivia {
+          try p.withSyntaxText(body: body)
+        }
+        try body(dat.tokenText)
+        for p in dat.trailingTrivia {
+          try p.withSyntaxText(body: body)
+        }
+      }
+    case .layout(let dat):
+      for case let child? in dat.layout {
+        try child.withEachSyntaxText(body: body)
+      }
+    }
+  }
+
+  /// Retrieve the syntax text as an array of bytes that models the input
+  /// source even in the presence of invalid UTF-8.
+  public var syntaxTextBytes: [UInt8] {
```

Comment:
Could we use `Data` instead of `[UInt8]`?

Reply:
No, we can't take a dependency on Foundation for anything that goes in a compiler. It's too large for any non-Darwin platform.

Reply:
And I assume that it’s alright to use Foundation in `swift-parser-test` because it’s only a testing utility?

Reply:
Yes.